clippy_lints/src/needless_question_mark.rs

```diff
@@ -0,0 +1,220 @@
+use rustc_errors::Applicability;
+use rustc_hir::def::{CtorKind, CtorOf, DefKind, Res};
+use rustc_hir::{Body, Expr, ExprKind, LangItem, MatchSource, QPath};
+use rustc_lint::{LateContext, LateLintPass, LintContext};
+use rustc_middle::ty::DefIdTree;
+use rustc_semver::RustcVersion;
+use rustc_session::{declare_tool_lint, impl_lint_pass};
+use rustc_span::sym;
+
+use crate::utils;
+use if_chain::if_chain;
+
+declare_clippy_lint! {
+    /// **What it does:**
+    /// Suggests alternatives for useless applications of `?` in terminating expressions
+    ///
+    /// **Why is this bad?** There's no reason to use ? to short-circuit when execution of the body will end there anyway.
+    ///
+    /// **Known problems:** None.
+    ///
+    /// **Example:**
+    ///
+    /// ```rust,ignore
+    /// struct Outer {
+    ///     inner: Option<String>
+    /// }
+    ///
+    /// let a = Some(Outer { inner: Some("hello".into()) });
+    /// let b = a.and_then(|a| Some(a.inner?));
+    ///
+    /// let c = Some(b.inner?);
+    /// let d = Ok(b.inner?);
```

Comment:
Is this even linted? I don't think this lint applies to assignments.

Reply:
Whoops, sorry about that. It'll be finished by today.